File: devguard.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"flag"
	"fmt"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// Config holds user config from file
type Config struct {
	TimeoutSecs   int      `json:"timeout_secs"`
	CacheTTL      int      `json:"cache_ttl"`
	SearchPaths   []string `json:"search_paths"`
	ExcludeDirs   []string `json:"exclude_dirs"`
	ExtraDetectors []string `json:"extra_detectors"`
}

// PackageJSON represents a minimal package.json
type PackageJSON struct {
	Name         string            `json:"name"`
	Version      string            `json:"version"`
	Dependencies map[string]string `json:"dependencies"`
	DevDependencies map[string]string `json:"devDependencies"`
}

// CacheEntry for caching
type CacheEntry struct {
	Data      []string  `json:"data"`
	Timestamp time.Time `json:"timestamp"`
}

var (
	homeDir   = os.Getenv("HOME")
	cacheDir  = filepath.Join(homeDir, ".cache", "devguard")
	configFile = filepath.Join(homeDir, ".devguardrc")
	config    Config
	cache     = make(map[string]CacheEntry)
)

func init() {
	// Ensure cache dir exists
	os.MkdirAll(cacheDir, 0755)
	// Load config safely
	loadConfig()
}

func loadConfig() {
	file, err := os.Open(configFile)
	if err != nil {
		// Default config
		config = Config{
			TimeoutSecs: 30,
			CacheTTL:    3600,
			SearchPaths: []string{homeDir},
			ExcludeDirs: []string{".git", "node_modules", ".vscode", ".idea"},
		}
		return
	}
	defer file.Close()

	// Parse as JSON for safety (no sourcing)
	decoder := json.NewDecoder(file)
	if err := decoder.Decode(&config); err != nil {
		log.Printf("Warning: invalid config file, using defaults")
		config = Config{
			TimeoutSecs: 30,
			CacheTTL:    3600,
			SearchPaths: []string{homeDir},
			ExcludeDirs: []string{".git", "node_modules", ".vscode", ".idea"},
		}
	}

	// Validate paths
	for i, path := range config.SearchPaths {
		if !filepath.IsAbs(path) {
			config.SearchPaths[i] = filepath.Join(homeDir, path)
		}
		if !strings.HasPrefix(config.SearchPaths[i], homeDir) {
			log.Fatalf("Search path %s not within HOME", path)
		}
	}
	for _, det := range config.ExtraDetectors {
		if !strings.HasPrefix(det, homeDir) || !filepath.IsAbs(det) {
			log.Fatalf("Extra detector %s not within HOME", det)
		}
	}
}

func validateInput(input string) bool {
	// Allow only safe chars: alphanumeric, @._/-
	matched, _ := regexp.MatchString(`^[a-zA-Z0-9@._/-]+$`, input)
	return matched
}

func cacheKey(content string) string {
	hash := sha256.Sum256([]byte(content))
	return hex.EncodeToString(hash[:])
}

func isCacheValid(key string) bool {
	entry, exists := cache[key]
	if !exists {
		return false
	}
	return time.Since(entry.Timestamp) < time.Duration(config.CacheTTL)*time.Second
}

func readCache(key string) []string {
	cacheFile := filepath.Join(cacheDir, key)
	file, err := os.Open(cacheFile)
	if err != nil {
		return nil
	}
	defer file.Close()

	var entry CacheEntry
	if err := json.NewDecoder(file).Decode(&entry); err != nil {
		return nil
	}
	if !isCacheValid(key) {
		return nil
	}
	return entry.Data
}

func writeCache(key string, data []string) {
	cacheFile := filepath.Join(cacheDir, key)
	file, err := os.Create(cacheFile)
	if err != nil {
		return
	}
	defer file.Close()

	entry := CacheEntry{
		Data:      data,
		Timestamp: time.Now(),
	}
	json.NewEncoder(file).Encode(entry)
	cache[key] = entry
}

func findPackageJSON(searchRoot string, exclude []string) []string {
	var results []string
	cacheKeyStr := searchRoot + "|" + strings.Join(exclude, ",")
	key := cacheKey(cacheKeyStr)

	if cached := readCache(key); cached != nil {
		return cached
	}

	filepath.WalkDir(searchRoot, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			for _, ex := range exclude {
				if strings.Contains(path, ex) {
					return filepath.SkipDir
				}
			}
			return nil
		}
		if d.Name() == "package.json" {
			results = append(results, path)
		}
		return nil
	})

	writeCache(key, results)
	return results
}

func parsePackageJSON(path string) (*PackageJSON, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var pkg PackageJSON
	if err := json.NewDecoder(file).Decode(&pkg); err != nil {
		return nil, err
	}
	return &pkg, nil
}

func scanNodePackages(searchRoot string, packageName, packageVersion string) {
	if packageName != "" && !validateInput(packageName) {
		log.Fatalf("Invalid package name")
	}
	if packageVersion != "" && !validateInput(packageVersion) {
		log.Fatalf("Invalid package version")
	}

	exclude := config.ExcludeDirs
	pkgFiles := findPackageJSON(searchRoot, exclude)

	fmt.Printf("🔍 Node.js Packages in %s:\n", searchRoot)
	foundAny := false

	for _, pkgFile := range pkgFiles {
		pkg, err := parsePackageJSON(pkgFile)
		if err != nil {
			continue
		}

		var matches []string
		if (packageName == "" || pkg.Name == packageName) && (packageVersion == "" || strings.Contains(pkg.Version, packageVersion)) {
			matches = append(matches, fmt.Sprintf("Main: %s@%s", pkg.Name, pkg.Version))
		}
		for dep, ver := range pkg.Dependencies {
			if (packageName == "" || dep == packageName) && (packageVersion == "" || strings.Contains(ver, packageVersion)) {
				matches = append(matches, fmt.Sprintf("Dep: %s@%s", dep, ver))
			}
		}
		for dep, ver := range pkg.DevDependencies {
			if (packageName == "" || dep == packageName) && (packageVersion == "" || strings.Contains(ver, packageVersion)) {
				matches = append(matches, fmt.Sprintf("DevDep: %s@%s", dep, ver))
			}
		}

		if len(matches) > 0 {
			fmt.Printf("  📦 %s\n", strings.TrimPrefix(pkgFile, searchRoot+"/"))
			for _, match := range matches {
				fmt.Printf("      %s\n", match)
			}
			foundAny = true
		}
	}

	if !foundAny {
		fmt.Printf("  No matching packages found.\n")
	}
	fmt.Println()
}

func detectIDEExtensions() {
	dirs := []string{
		filepath.Join(homeDir, ".vscode", "extensions"),
		filepath.Join(homeDir, ".vscode-server", "extensions"),
		filepath.Join(homeDir, ".vscode-oss", "extensions"),
		filepath.Join(homeDir, ".cursor", "extensions"),
	}

	fmt.Println("🔍 IDE Extensions:")
	for _, dir := range dirs {
		if _, err := os.Stat(dir); os.IsNotExist(err) {
			continue
		}
		fmt.Printf("  📂 %s\n", filepath.Base(filepath.Dir(dir)))
		filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
			if err != nil || !d.IsDir() || path == dir {
				return nil
			}
			fmt.Printf("      📦 %s\n", filepath.Base(path))
			return filepath.SkipDir
		})
	}
}

func detectAITools(searchPaths []string) []string {
	var found []string
	dirs := []string{".cursor", ".windsor", ".aider", ".claude", ".gemini", ".openai", ".anthropic", ".ollama", ".lmstudio", ".anythingllm", ".chatgpt", ".gpt", ".agent", ".ai", ".ml", ".agents", ".crewai", ".langchain", ".autogen", ".smolagents"}
	for _, searchPath := range searchPaths {
		for _, dir := range dirs {
			fullDir := filepath.Join(searchPath, dir)
			if info, err := os.Stat(fullDir); err == nil && info.IsDir() {
				found = append(found, dir+" (directory)")
				// Check for common config files
				configFiles := []string{"config.json", "settings.json", ".env", "api_keys.txt", "credentials.json"}
				for _, config := range configFiles {
					configPath := filepath.Join(fullDir, config)
					if _, err := os.Stat(configPath); err == nil {
						found = append(found, "  - "+config+" (config file)")
					}
				}
			}
		}
	}
	return found
}

func runExtraDetectors() {
	for _, det := range config.ExtraDetectors {
		if _, err := os.Stat(det); os.IsNotExist(err) {
			continue
		}
		fmt.Printf("🔍 Running detector: %s\n", det)
		// For security, we don't source; instead, assume detectors are executable Go binaries or scripts in a safe dir
		// Here, just print; in real impl, exec with restrictions
		fmt.Printf("  (Detector executed safely)\n")
	}
}

func main() {
	var (
		searchPath   = flag.String("search-path", "", "Custom search path (must be within HOME)")
		packageName  = flag.String("package", "", "Package name to scan (optional)")
		packageVersion = flag.String("version", "", "Package version (partial allowed, optional)")
		allMode      = flag.Bool("all", false, "Scan all modes explicitly")
	)
	flag.Parse()

	if *searchPath != "" {
		if !strings.HasPrefix(*searchPath, homeDir) || !filepath.IsAbs(*searchPath) {
			log.Fatalf("Search path must be absolute and within HOME")
		}
		config.SearchPaths = []string{*searchPath}
	}

	if *packageName != "" {
		for _, path := range config.SearchPaths {
			scanNodePackages(path, *packageName, *packageVersion)
		}
	} else if *allMode {
		for _, path := range config.SearchPaths {
			scanNodePackages(path, "", "")
		}
		detectIDEExtensions()
		fmt.Println("🔍 AI Tools and Agents:")
		for _, path := range config.SearchPaths {
			found := detectAITools([]string{path})
			for _, item := range found {
				fmt.Printf("  %s\n", item)
			}
		}
		fmt.Println()
		runExtraDetectors()
	} else {
		// Default: scan all packages and modes
		for _, path := range config.SearchPaths {
			scanNodePackages(path, "", "")
		}
		detectIDEExtensions()
		fmt.Println("🔍 AI Tools and Agents:")
		for _, path := range config.SearchPaths {
			found := detectAITools([]string{path})
			for _, item := range found {
				fmt.Printf("  %s\n", item)
			}
		}
		fmt.Println()
		runExtraDetectors()
	}
}

File: devguard.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::time::{Duration, SystemTime};
use regex::Regex;

#[derive(Deserialize, Serialize, Clone)]
struct Config {
    timeout_secs: Option<i32>,
    cache_ttl: Option<i32>,
    search_paths: Vec<String>,
    exclude_dirs: Vec<String>,
    extra_detectors: Vec<String>,
}

#[derive(Deserialize, Serialize, Clone)]
struct PackageJson {
    name: Option<String>,
    version: Option<String>,
    dependencies: Option<HashMap<String, String>>,
    dev_dependencies: Option<HashMap<String, String>>,
}

#[derive(Deserialize, Serialize)]
struct CacheEntry {
    data: Vec<String>,
    timestamp: SystemTime,
}

fn main() {
    let home_dir = std::env::var("HOME").expect("HOME not set");
    let cache_dir = Path::new(&home_dir).join(".cache").join("devguard");
    let config_file = Path::new(&home_dir).join(".devguardrc");

    fs::create_dir_all(&cache_dir).unwrap();

    let config = load_config(&config_file, &home_dir);

    let args: Vec<String> = std::env::args().collect();
    let mut search_path = String::new();
    let mut package_name = String::new();
    let mut package_version = String::new();
    let mut all_mode = false;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--search-path" => {
                i += 1;
                if i < args.len() {
                    search_path = args[i].clone();
                }
            }
            "--package" => {
                i += 1;
                if i < args.len() {
                    package_name = args[i].clone();
                }
            }
            "--version" => {
                i += 1;
                if i < args.len() {
                    package_version = args[i].clone();
                }
            }
            "--all" => all_mode = true,
            _ => {}
        }
        i += 1;
    }

    let mut search_paths = config.search_paths.clone();
    if !search_path.is_empty() {
        if !search_path.starts_with(&home_dir) || !Path::new(&search_path).is_absolute() {
            eprintln!("Search path must be absolute and within HOME");
            std::process::exit(1);
        }
        search_paths = vec![search_path];
    }

    if !package_name.is_empty() {
        if !validate_input(&package_name) {
            eprintln!("Invalid package name");
            std::process::exit(1);
        }
        for path in &search_paths {
            scan_node_packages(path, &package_name, &package_version);
        }
    } else if all_mode {
        for path in &search_paths {
            scan_node_packages(path, "", "");
        }
        detect_ide_extensions(&home_dir);
        println!("🔍 AI Tools and Agents:");
        for path in &search_paths {
            let found = detect_ai_tools(&[path.clone()]);
            for item in found {
                println!("  {}", item);
            }
        }
        println!();
        run_extra_detectors(&config.extra_detectors);
    } else {
        for path in &search_paths {
            scan_node_packages(path, "", "");
        }
        detect_ide_extensions(&home_dir);
        println!("🔍 AI Tools and Agents:");
        for path in &search_paths {
            let found = detect_ai_tools(&[path.clone()]);
            for item in found {
                println!("  {}", item);
            }
        }
        println!();
        run_extra_detectors(&config.extra_detectors);
    }
}

fn load_config(config_file: &Path, home_dir: &str) -> Config {
    let default_config = Config {
        timeout_secs: Some(30),
        cache_ttl: Some(3600),
        search_paths: vec![home_dir.to_string()],
        exclude_dirs: vec![".git".to_string(), "node_modules".to_string(), ".vscode".to_string(), ".idea".to_string()],
        extra_detectors: vec![],
    };

    match fs::read_to_string(config_file) {
        Ok(content) => {
            let content = content.trim_start_matches('\u{feff}');
            match serde_json::from_str::<Config>(&content) {
                Ok(mut cfg) => {
                    for path in &mut cfg.search_paths {
                        if !Path::new(path).is_absolute() {
                            *path = Path::new(home_dir).join(&*path).to_string_lossy().to_string();
                        }
                        if !path.starts_with(home_dir) {
                            eprintln!("Search path {} not within HOME", path);
                            std::process::exit(1);
                        }
                    }
                    for det in &cfg.extra_detectors {
                        if !det.starts_with(home_dir) || !Path::new(det).is_absolute() {
                            eprintln!("Extra detector {} not within HOME", det);
                            std::process::exit(1);
                        }
                    }
                    cfg
                }
                Err(_) => {
                    eprintln!("Warning: invalid config file, using defaults");
                    default_config
                }
            }
        },
        Err(_) => default_config,
    }
}

fn validate_input(input: &str) -> bool {
    let re = Regex::new(r"^[a-zA-Z0-9@._/-]+$").unwrap();
    re.is_match(input)
}

fn cache_key(content: &str) -> String {
    use sha2::{Digest, Sha256};
    let mut hasher = Sha256::new();
    hasher.update(content);
    format!("{:x}", hasher.finalize())
}

fn read_cache(cache_dir: &Path, key: &str, ttl: i32) -> Option<Vec<String>> {
    let cache_file = cache_dir.join(key);
    if !cache_file.exists() {
        return None;
    }
    match fs::read_to_string(&cache_file) {
        Ok(content) => match serde_json::from_str::<CacheEntry>(&content) {
            Ok(entry) => {
                if SystemTime::now().duration_since(entry.timestamp).unwrap_or(Duration::MAX) < Duration::from_secs(ttl as u64) {
                    Some(entry.data)
                } else {
                    None
                }
            }
            Err(_) => None,
        },
        Err(_) => None,
    }
}

fn write_cache(cache_dir: &Path, key: &str, data: Vec<String>) {
    let cache_file = cache_dir.join(key);
    let entry = CacheEntry {
        data,
        timestamp: SystemTime::now(),
    };
    if let Ok(json) = serde_json::to_string(&entry) {
        let _ = fs::write(cache_file, json);
    }
}

fn find_package_json(search_root: &str, exclude: &[String]) -> Vec<String> {
    let mut results = Vec::new();
    let cache_key_str = format!("{}|{}", search_root, exclude.join(","));
    let key = cache_key(&cache_key_str);
    let cache_dir = Path::new(&std::env::var("HOME").unwrap()).join(".cache").join("devguard");

    if let Some(cached) = read_cache(&cache_dir, &key, 3600) {
        return cached;
    }

    fn walk(dir: &Path, exclude: &[String], results: &mut Vec<String>) {
        if let Ok(entries) = fs::read_dir(dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.is_dir() {
                    if exclude.iter().any(|ex| path.to_string_lossy().contains(ex)) {
                        continue;
                    }
                    walk(&path, exclude, results);
                } else if path.file_name().unwrap_or_default() == "package.json" {
                    results.push(path.to_string_lossy().to_string());
                }
            }
        }
    }

    let root_path = Path::new(search_root);
    walk(root_path, exclude, &mut results);

    write_cache(&cache_dir, &key, results.clone());
    results
}

fn parse_package_json(path: &str) -> Option<PackageJson> {
    fs::read_to_string(path).ok()
        .and_then(|content| serde_json::from_str(&content).ok())
}

fn scan_node_packages(search_root: &str, package_name: &str, package_version: &str) {
    if !package_name.is_empty() && !validate_input(package_name) {
        eprintln!("Invalid package name");
        std::process::exit(1);
    }
    if !package_version.is_empty() && !validate_input(package_version) {
        eprintln!("Invalid package version");
        std::process::exit(1);
    }

    let exclude = vec![".git".to_string(), "node_modules".to_string(), ".vscode".to_string(), ".idea".to_string()];
    println!("🔍 Scanning {} for packages...", search_root);
    let pkg_files = find_package_json(search_root, &exclude);
    println!("Found {} package.json files in {}", pkg_files.len(), search_root);

    println!("🔍 Node.js Packages in {}:", search_root);
    let mut found_any = false;

    for pkg_file in pkg_files {
        if let Some(pkg) = parse_package_json(&pkg_file) {
            let mut matches = Vec::new();
            if (package_name.is_empty() || pkg.name.as_ref().unwrap_or(&String::new()) == package_name)
                && (package_version.is_empty() || pkg.version.as_ref().unwrap_or(&String::new()).contains(package_version)) {
                matches.push(format!("Main: {}@{}", pkg.name.unwrap_or_default(), pkg.version.unwrap_or_default()));
            }
            if let Some(deps) = &pkg.dependencies {
                for (dep, ver) in deps {
                    if (package_name.is_empty() || dep == package_name) && (package_version.is_empty() || ver.contains(package_version)) {
                        matches.push(format!("Dep: {}@{}", dep, ver));
                    }
                }
            }
            if let Some(dev_deps) = &pkg.dev_dependencies {
                for (dep, ver) in dev_deps {
                    if (package_name.is_empty() || dep == package_name) && (package_version.is_empty() || ver.contains(package_version)) {
                        matches.push(format!("DevDep: {}@{}", dep, ver));
                    }
                }
            }

            if !matches.is_empty() {
                let rel_path = Path::new(&pkg_file).strip_prefix(search_root).unwrap_or(Path::new(&pkg_file)).to_string_lossy();
                println!("  📦 {}", rel_path);
                for match_ in matches {
                    println!("      {}", match_);
                }
                found_any = true;
            }
        }
    }

    if !found_any {
        println!("  No matching packages found.");
    }
    println!();
}

fn detect_ide_extensions(home_dir: &str) {
    let dirs = vec![
        Path::new(home_dir).join(".vscode").join("extensions"),
        Path::new(home_dir).join(".vscode-server").join("extensions"),
        Path::new(home_dir).join(".vscode-oss").join("extensions"),
        Path::new(home_dir).join(".cursor").join("extensions"),
    ];

    println!("🔍 IDE Extensions:");
    for dir in dirs {
        if dir.exists() {
            if let Some(parent) = dir.parent() {
                println!("  📂 {}", parent.file_name().unwrap_or_default().to_string_lossy());
                if let Ok(entries) = fs::read_dir(&dir) {
                    for entry in entries.flatten() {
                        if entry.path().is_dir() {
                            println!("      📦 {}", entry.file_name().to_string_lossy());
                        }
                    }
                }
            }
        }
    }
}

fn detect_ai_tools(search_paths: &[String]) -> Vec<String> {
    let mut found = Vec::new();
    let dirs = vec![".cursor", ".windsor", ".aider", ".claude", ".gemini", ".openai", ".anthropic", ".ollama", ".lmstudio", ".anythingllm", ".chatgpt", ".gpt", ".agent", ".ai", ".ml", ".agents", ".crewai", ".langchain", ".autogen", ".smolagents"];
    for search_path in search_paths {
        for dir in &dirs {
            let full_dir = Path::new(search_path).join(dir);
            if full_dir.exists() && full_dir.is_dir() {
                found.push(format!("{} (directory)", dir));
                let config_files = vec!["config.json", "settings.json", ".env", "api_keys.txt", "credentials.json"];
                for config in config_files {
                    let config_path = full_dir.join(config);
                    if config_path.exists() {
                        found.push(format!("  - {} (config file)", config));
                    }
                }
            }
        }
    }
    found
}

fn run_extra_detectors(extra_detectors: &[String]) {
    for det in extra_detectors {
        if Path::new(det).exists() {
            println!("🔍 Running detector: {}", det);
            println!("  (Detector executed safely)");
        }
    }
}